### PR TITLE
ISD-2100 Refactor tls_relation module and charm state logic

### PR DESCRIPTION
**Pull request by Thanhphan1147**
Follow up to comments on #28 and internal discussions. The `tls_relation` module needs to be simplified and the `certificates` integration handling logic needs to be improved.

This PR refactors and tries to simplify the `tls_relation` module, the `tls_information` charm state component and the `SecretResourceDefinition` used to manage k8s secrets:

* `tls_relation` module is now initialized with the certificates requirer v3 lib, replacing all direct modification of the integration data with library calls. Remove useless code that stores the certificates information on the application's integration data.
* `TLSInformation` state component now no longer stores the tls integration integration.
* `SecretResourceDefinition` now is built directly from TLSInformation using `SecretResourceDefinition.from_tls_information` class method
* Simplify the `certificates` integration handling logic.
* Update tests to accomodate the above changes

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

**Comment by Thanhphan1147** [comment]
Hi @gregory-schiano, can you please take a look at this PR and leave your review? Thanks!

File: src/charm.py [hunk 1]
```diff
@@ -177,116 +173,55 @@ def _on_get_certificate_action(self, event: ActionEvent) -> None:
...
-            return
-
-        event.set_results(
-            {
-                f"certificate-{hostname}": tls_rel_data[f"certificate-{hostname}"],
-                f"ca-{hostname}": tls_rel_data[f"ca-{hostname}"],
-                f"chain-{hostname}": tls_rel_data[f"chain-{hostname}"],
-            }
-        )
+        TLSInformation.from_charm(self, self.certificates)
```
Comment (cbartz) [inline src/charm.py:176]: This line is not obvious, I assume there are relevant side-effects in it, but I think there should at least be a comment pointing this out (because no return value is used and the name of the function does not explain the side-effects). It would be better if we could have a solution without relying on side-effects, as it is easier to reason about.

File: src/charm.py [hunk 2]
```diff
@@ -177,116 +173,55 @@ def _on_get_certificate_action(self, event: ActionEvent) -> None:
...
+                return
+
+        event.fail(f"Missing or incomplete certificate data for {hostname}")
 
     @validate_config_and_integration(defer=True)
     def _on_certificates_relation_created(self, _: RelationCreatedEvent) -> None:
         """Handle the TLS Certificate relation created event."""
-        tls_information = TLSInformation.from_charm(self)
-
+        TLSInformation.from_charm(self, self.certificates)
```
Comment (cbartz) [inline src/charm.py:194]: same here

File: src/charm.py [hunk 3]
```diff
@@ -177,116 +173,55 @@ def _on_get_certificate_action(self, event: ActionEvent) -> None:
...
-            config.external_hostname, tls_information.tls_requirer_integration
-        )
+        self._tls.certificate_relation_created(config.external_hostname)
 
     @validate_config_and_integration(defer=True)
     def _on_certificates_relation_joined(self, _: RelationJoinedEvent) -> None:
         """Handle the TLS Certificate relation joined event."""
-        tls_information = TLSInformation.from_charm(self)
-
+        TLSInformation.from_charm(self, self.certificates)
```
Comment (cbartz) [inline src/charm.py:202]: same here

File: src/charm.py [hunk 4]
```diff
@@ -177,116 +173,55 @@ def _on_get_certificate_action(self, event: ActionEvent) -> None:
...
-                self._tls.pop_relation_data_fields(
-                    [f"key-{hostname}", f"password-{hostname}"],
-                    tls_requirer_integration,  # type: ignore[arg-type]
-                )
-            except KeyError:
-                logger.warning("Relation data for %s already does not exist", hostname)
-            self.certificates.request_certificate_revocation(
-                certificate_signing_request=old_csr.encode()
-            )
+        TLSInformation.from_charm(self, self.certificates)
```
Comment (cbartz) [inline src/charm.py:223]: same here

File: src/charm.py [hunk 5]
```diff
@@ -295,41 +230,26 @@ def _on_certificate_invalidated(self, event: CertificateInvalidatedEvent) -> Non
...
-            tls_information = TLSInformation.from_charm(self)
-        except (TlsIntegrationMissingError, InvalidCharmConfigError) as exc:
-            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
-            logger.error(
-                "(%s) charm is not ready to handle this event, deferring: %s.", event, exc
-            )
-            event.defer()
-            return
-
+        TLSInformation.from_charm(self, self.certificates)
```
Comment (cbartz) [inline src/charm.py:233]: same here

File: src/charm.py [hunk 6]
```diff
@@ -295,41 +230,26 @@ def _on_certificate_invalidated(self, event: CertificateInvalidatedEvent) -> Non
...
+            self._tls.certificate_expiring(event)
         self.unit.status = MaintenanceStatus("Waiting for new certificate")
 
     @validate_config_and_integration(defer=True)
     def _on_all_certificates_invalidated(self, _: AllCertificatesInvalidatedEvent) -> None:
         """Handle the TLS Certificate relation broken event."""
-        tls_information = TLSInformation.from_charm(self)
-        tls_information.tls_requirer_integration.data[self.app].clear()
-
+        TLSInformation.from_charm(self, self.certificates)
```
Comment (cbartz) [inline src/charm.py:243]: same here

File: src/resource_manager/gateway.py [hunk 7]
```diff
@@ -92,38 +92,39 @@ def _label_selector(self) -> str:
         return ",".join(f"{k}={v}" for k, v in self._labels.items())
 
     @map_k8s_auth_exception
-    def _gen_resource(self, state: ResourceDefinition) -> dict:
+    def _gen_resource(self, resource_definition: ResourceDefinition) -> dict:
         """Generate a Gateway resource from a gateway resource definition.
 
         Args:
-            state: Part of charm state.
+            resource_definition: Part of charm state.
```
Comment (cbartz) [inline src/resource_manager/gateway.py:99]: Can you rephrase this description?

File: src/resource_manager/resource_manager.py [hunk 8]
```diff
@@ -45,11 +45,11 @@ class ResourceManager(typing.Protocol[AnyResource]):
     """Abstract base class for a generic Kubernetes resource controller."""
 
     @abc.abstractmethod
-    def _gen_resource(self, state: ResourceDefinition) -> AnyResource:
+    def _gen_resource(self, resource_definition: ResourceDefinition) -> AnyResource:
         """Abstract method to generate a resource from ingress definition.
 
         Args:
-            state: Part of charm state consisting of one or several components.
+            resource_definition: Part of charm state consisting of one or several components.
```
Comment (cbartz) [inline src/resource_manager/resource_manager.py:52]: Can you rephrase this description?

File: src/resource_manager/secret.py [hunk 9]
```diff
@@ -25,43 +23,54 @@
...
+            hostname: The requested hostname.
+
+        Raises:
+            CertificateDataNotReadyError: When the certificate data is not ready.
+
+        Returns:
+            SecretResourceDefinition: Information about the certificate.
         """
-        super().__init__(gateway_resource_information, charm_config, tls_information)
+        if hostname not in tls_information.tls_certs.keys():
```
Comment (cbartz) [inline src/resource_manager/secret.py:64]:
```suggestion
        if hostname not in tls_information.tls_certs:
```

if its a dict , using `in` without `keys()` should suffice

File: tests/unit/conftest.py [hunk 10]
```diff
@@ -95,3 +95,13 @@ def mock_certificate_fixture() -> str:
         "4+3+0/6Ba2Zlt9fu4PixG+XukQnBIxtIMjWp7q7xWp8F4aOW"
         "-----END CERTIFICATE-----"
     )
+    provider_cert_mock = MagicMock()
```
Comment (cbartz) [inline tests/unit/conftest.py:98]: consider using the `spec` attribute for MagicMock

**Review by cbartz — COMMENTED** [review]
Thanks for the refactoring! I think you forgot to do src-docs. I think there are also missing unit tests, but I guess you are preparing a PR for those?

File: src/charm.py (hunk 1 above)
Comment (gregory-schiano) [reply src/charm.py:176]: True, I missed that one

File: src/state/tls.py [hunk 11]
```diff
@@ -22,29 +30,36 @@ class TLSInformation:
...
             TlsIntegrationMissingError: When integration is not ready.
+            SecretNotSupportedError: When running with incompatible juju version.
 
         Returns:
             TLSInformation: Information about configured TLS certs.
         """
+        if not JujuVersion.from_environ().has_secrets:
+            raise SecretNotSupportedError(
+                "The charm requires the 'secrets' feature to be supported (juju version >= 3.0.3)."
+            )
```
Comment (gregory-schiano) [inline src/state/tls.py:61]: I don't think this is necessary since we set:
https://github.com/canonical/gateway-api-integrator-operator/blob/52b18b5c0a5f1c5002d9977854b815f6aa491ae8/charmcraft.yaml#L36

Meaning that Juju won't let the charm being deployed if the controller is not >3.3

File: src/tls_relation.py [hunk 12]
```diff
@@ -5,35 +5,54 @@
...
+        certificate: The certificate in PEM format.
+
+    Returns:
+        The hostname the certificate is issue to.
+    """
+    decoded_cert = x509.load_pem_x509_certificate(certificate.encode())
+
+    common_name_attribute = decoded_cert.subject.get_attributes_for_oid(NameOID.COMMON_NAME)
+    if not common_name_attribute:
+        return ""
```
Comment (gregory-schiano) [inline src/tls_relation.py:38]: Not sure we should return an empty string.
If you think we should, then when you call that method you should be defensive

File: src/tls_relation.py [hunk 13]
```diff
@@ -5,35 +5,54 @@
...
         """Init method for the class.
 
         Args:
-            model: The charm model used to get the relations and secrets.
+            certificates: The TLS certificates requirer library.
         """
-        self.charm_model = model
-        self.charm_app = model.app
+        self.certificates = certificates
+        self.model = self.certificates.model
```
Comment (gregory-schiano) [inline src/tls_relation.py:53]: It's a weird way to access the model, very indirect. Why removing the model from the constructor ?

File: src/tls_relation.py [hunk 14]
```diff
@@ -44,161 +63,87 @@ def generate_password(self) -> str:
...
 
-    def certificate_relation_created(self, hostname: str, tls_integration: Relation) -> None:
+    def certificate_relation_created(self, hostname: str) -> None:
         """Handle the TLS Certificate created event.
 
         Args:
             hostname: Certificate's hostname.
-            tls_integration: The tls certificates integration.
         """
+        # At this point, TLSInformation state component should already be initialized
```
Comment (gregory-schiano) [inline src/tls_relation.py:87]: Perhaps adding an assert before casting ?

File: src/tls_relation.py [hunk 15]
```diff
@@ -207,9 +152,20 @@ def _get_private_key(self, hostname: str) -> Dict[str, str]:
...
-        private_key_dict = {}
-        if JujuVersion.from_environ().has_secrets:
-            secret = self.charm_model.get_secret(label=f"private-key-{hostname}")
-            private_key_dict["key"] = secret.get_content()["key"]
-            private_key_dict["password"] = secret.get_content()["password"]
-        return private_key_dict
+        secret = self.model.get_secret(label=f"private-key-{hostname}")
+        private_key = secret.get_content()["key"]
+        password = secret.get_content()["password"]
+        return (private_key, password)
```
Comment (gregory-schiano) [inline src/tls_relation.py:158]: I think you could create a `KeyPair` Namedtuple as this is often used at many places sometimes as a dict sometimes as a tuple

Comment (Thanhphan1147) [reply src/tls_relation.py]: Makes sense, I updated the method to return a `KeyPair` named tuple.

File: src/tls_relation.py (hunk 14 above)
Comment (Thanhphan1147) [reply src/tls_relation.py:109]: I added an assert here.

File: src/tls_relation.py (hunk 13 above)
Comment (Thanhphan1147) [reply src/tls_relation.py]: I was aiming to reduce the number of dependencies for the initialization of this module. I added back the `model` parameter.

File: src/tls_relation.py (hunk 12 above)
Comment (Thanhphan1147) [reply src/tls_relation.py]: I think the idea behind this was that we can trust that the provider will send us valid certificates so we don't need to be defensive. I've thought about it for a while and I agree that doing it like this is a bit fragile since now the method is used in way more places than in the `nginx-ingress-integrator` charm and the output of this is used to lookup secrets and re-requesting new certs as well so raising an exception here makes sense for me. I've add that code in and handle the thrown exception in `validation.py`

File: src/charm.py (hunk 1 above)
Comment (Thanhphan1147) [reply src/charm.py]: we are replying on the `from_charm` method to check for the existence of the `certificates` integration and raises an exception which is handled by the `@validate_config_and_integration` decorator. I've added another class method called `validate` and used it here to be more explicit on what we're trying to do.

File: src/charm.py (hunk 3 above)
Comment (Thanhphan1147) [reply src/charm.py]: we are replying on the `from_charm` method to check for the existence of the `certificates` integration and raises an exception which is handled by the `@validate_config_and_integration` decorator. I've added another class method called `validate` and used it here to be more explicit on what we're trying to do.

File: src/charm.py (hunk 2 above)
Comment (Thanhphan1147) [reply src/charm.py]: we are replying on the `from_charm` method to check for the existence of the `certificates` integration and raises an exception which is handled by the `@validate_config_and_integration` decorator. I've added another class method called `validate` and used it here to be more explicit on what we're trying to do.

File: src/charm.py (hunk 4 above)
Comment (Thanhphan1147) [reply src/charm.py]: Same comments above

File: src/charm.py (hunk 5 above)
Comment (Thanhphan1147) [reply src/charm.py]: Same comments above

File: src/charm.py (hunk 6 above)
Comment (Thanhphan1147) [reply src/charm.py]: Same comments above

File: src/resource_manager/gateway.py (hunk 7 above)
Comment (Thanhphan1147) [reply src/resource_manager/gateway.py]: I rephrased this to be more explicit.

File: tests/unit/conftest.py (hunk 10 above)
Comment (Thanhphan1147) [reply tests/unit/conftest.py:98]: I rephrased this to be more explicit.

File: src/tls_relation.py [hunk 16]
```diff
@@ -5,200 +5,169 @@
...
-            model: The charm model used to get the relations and secrets.
-        """
-        self.charm_model = model
-        self.charm_app = model.app
+class KeyPair(typing.NamedTuple):
+    """Stores a private key and encryption password.
 
-    def generate_password(self) -> str:
-        """Generate a random 12 character password.
+    Attrs:
```
Comment (yanksyoon) [inline src/tls_relation.py:32]: I think `Attrs` doesn't work with docstrings-complete. `Attributes` should work though :D

**Review by yanksyoon — DISMISSED** [review]
LGTM! Minor nits :D

File: src/state/tls.py [hunk 17]
```diff
@@ -22,55 +25,61 @@ class TLSInformation:
...
         tls_certs = {}
         tls_keys = {}
         secret_resource_name_prefix = f"{charm.app.name}-secret"
 
-        for key, value in tls_requirer_integration.data[charm.app].items():
-            if key.startswith("chain-"):
-                hostname = key.split("-", maxsplit=1)[1]
-                tls_certs[hostname] = value
-
-                if JujuVersion.from_environ().has_secrets:
```
[the remark this answers is not in the document]
Comment (gregory-schiano) [reply src/state/tls.py:64]: That's already set to juju > 3.3

File: src/tls_relation.py (hunk 16 above)
Comment (cbartz) [reply src/tls_relation.py:32]: src-docs doesn't render it properly if its named `Attrs`

File: src/tls_relation.py [hunk 18]
```diff
@@ -85,7 +107,9 @@ def certificate_relation_created(self, hostname: str) -> None:
             hostname: Certificate's hostname.
         """
         # At this point, TLSInformation state component should already be initialized
-        tls_integration = typing.cast(Relation, self.model.get_relation(self.integration_name))
+        tls_integration = self.model.get_relation(self.integration_name)
+        assert tls_integration
```
Comment (cbartz) [inline src/tls_relation.py:111]: this will probably lead to a bandit warning once linting is enabled: https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html

**Review by cbartz — DISMISSED** [review]

File: src/tls_relation.py (hunk 18 above)
Comment (gregory-schiano) [reply src/tls_relation.py:111]: We can use `raise AssertionError` instead

File: src/state/tls.py [hunk 19]
```diff
@@ -22,55 +25,61 @@ class TLSInformation:
...
 
         return cls(
             secret_resource_name_prefix=secret_resource_name_prefix,
-            tls_requirer_integration=tls_requirer_integration,
             tls_certs=tls_certs,
             tls_keys=tls_keys,
         )
+
+    @classmethod
+    def validate(cls, charm: ops.CharmBase) -> None:
```
[the remark this answers is not in the document]
Comment (gregory-schiano) [reply src/state/tls.py:71]: As discussed, I'm much more in favor of validation error to be exceptions than returning `None` that will require the caller to take a defensive approach

File: src/resource_manager/resource_manager.py (hunk 8 above)
Comment (Thanhphan1147) [reply src/resource_manager/resource_manager.py]: I rephrased this to be more explicit.

File: src/tls_relation.py (hunk 16 above)
Comment (Thanhphan1147) [reply src/tls_relation.py]: I renamed all `Attrs` to `Attributes`

File: src/tls_relation.py (hunk 18 above)
Comment (Thanhphan1147) [reply src/tls_relation.py]: Thanks, I updated the code to raise AssertionError.

**Review by cbartz — DISMISSED** [review]

**Review by cbartz — APPROVED** [review]

**Review by gregory-schiano — APPROVED** [review]
LGTM
